crates/sui-graphql-rpc/src/types/dry_run_result.rs

```diff
@@ -0,0 +1,153 @@
+
+    #[graphql(skip)]
+    pub events: Vec<NativeEvent>,
+
+    #[graphql(skip)]
+    pub effects: NativeTransactionEffects,
+
+    #[graphql(skip)]
+    pub tx_data: NativeTransactionData,
+}
+
+#[ComplexObject]
+impl DryRunResult {
+    /// The transaction block representing the dry run execution.
+    pub async fn transaction(&self) -> Option<TransactionBlock> {
+        Some(TransactionBlock::DryRun {
+            tx_data: self.tx_data.clone(),
+            effects: self.effects.clone(),
+            events: self.events.clone(),
+        })
+    }
+}
+
+impl TryFrom<DevInspectResults> for DryRunResult {
+    type Error = crate::error::Error;
+    fn try_from(results: DevInspectResults) -> Result<Self, Self::Error> {
+        let execution_results = results
+            .results
+            .ok_or(Error::Internal(
+                "No execution results returned from dev inspect".to_string(),
+            ))?
+            .into_iter()
+            .map(DryRunEffect::try_from)
+            .collect::<Result<Vec<_>, Self::Error>>()?;
+        let events = results
+            .events
+            .data
+            .into_iter()
+            .map(|e| NativeEvent {
+                sender: e.sender,
+                package_id: e.package_id,
+                transaction_module: e.transaction_module,
+                type_: e.type_,
+                contents: e.bcs,
+            })
+            .collect();
+        let effects: NativeTransactionEffects =
+            bcs::from_bytes(&results.raw_effects).map_err(|e| {
+                Error::Internal(format!("Unable to deserialize transaction effects: {e}"))
+            })?;
+        let tx_data: NativeTransactionData = bcs::from_bytes(&results.raw_txn_data)
+            .map_err(|e| Error::Internal(format!("Unable to deserialize transaction data: {e}")))?;
+        Ok(Self {
+            error: results.error,
+            results: Some(execution_results),
+            events,
+            effects,
+            tx_data,
+        })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, SimpleObject)]
+pub(crate) struct DryRunEffect {
+    /// Changes made to arguments that were mutably borrowed by each command in this transaction.
+    pub mutated_references: Option<Vec<DryRunMutation>>,
+
+    /// Return results of each command in this transaction.
+    pub return_values: Option<Vec<DryRunReturn>>,
+}
+
+impl TryFrom<SuiExecutionResult> for DryRunEffect {
+    type Error = crate::error::Error;
+
+    fn try_from(result: SuiExecutionResult) -> Result<Self, Self::Error> {
+        let mutated_references = result
+            .mutable_reference_outputs
+            .iter()
+            .map(|(argument, bcs, type_)| {
+                Ok(DryRunMutation {
+                    input: (*argument).into(),
+                    type_: MoveType::new(type_.clone().try_into()?),
+                    bcs: bcs.into(),
+                })
+            })
+            .collect::<Result<Vec<_>>>()
```

Comment:
I don't think you want to use the single parameter `Result` here -- that would turn the error into an `async-graphql` error -- this is what I was talking about before, about `use`-ing `crate::error::Error` -- it's to avoid situations like this.

Reply:
The error here actually isn't a `crate::error::Error` since it happens in json rpc type land. So I think the conversion to crate's Error is needed anyway?

Reply:
Ah got it -- I think we still don't want to use `Result<_>` though because this is `async_graphql::Result<_>` which means we are going from the JSON RPC error, to an `async_graphql::Error`, to a `crate::error::Error` and then back to an `async_graphql::Error`.